flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -521,8 +522,18 @@ public void acknowledgeCheckpoint(
             final ExecutionAttemptID executionAttemptID,
             final long checkpointId,
             final CheckpointMetrics checkpointMetrics,
-            final TaskStateSnapshot checkpointState) {
+            @Nullable final SerializedValue<TaskStateSnapshot> serializedCheckpointState) {
 
+        TaskStateSnapshot checkpointState;
+        try {
+            checkpointState =
+                    serializedCheckpointState == null
+                            ? null
+                            : serializedCheckpointState.deserializeValue(
+                                    ClassLoader.getSystemClassLoader());
```

Comment:
```suggestion
                                    getClass().getClassLoader());
```
It may be problematic for certain use-cases to assume that Flink is loaded by the system classloader.